process/block/shardblockProposal.go

```diff
@@ -97,6 +99,144 @@ func (sp *shardProcessor) CreateBlockProposal(
 	return shardHdr, body, nil
 }
 
+// VerifyBlockProposal verifies the proposed block. It returns nil if all ok or the specific error
+func (sp *shardProcessor) VerifyBlockProposal(
+	headerHandler data.HeaderHandler,
+	bodyHandler data.BodyHandler,
+	haveTime func() time.Duration,
+) error {
+	log.Debug("started verifying proposed block",
+		"epoch", headerHandler.GetEpoch(),
+		"shard", headerHandler.GetShardID(),
+		"round", headerHandler.GetRound(),
+		"nonce", headerHandler.GetNonce(),
+	)
+
+	err := sp.checkBlockValidity(headerHandler, bodyHandler)
+	if err != nil {
+		if errors.Is(err, process.ErrBlockHashDoesNotMatch) {
+			log.Debug("requested missing shard header",
+				"hash", headerHandler.GetPrevHash(),
+				"for shard", headerHandler.GetShardID(),
+			)
+
+			go sp.requestHandler.RequestShardHeaderForEpoch(headerHandler.GetShardID(), headerHandler.GetPrevHash(), headerHandler.GetEpoch())
```

Comment:
there might be a case here where the prev hash is from `headerHandler.GetEpoch() - 1`, so data won't be received